Implement Iceberg Kafka Connect with Delta Writer Support in DV Mode for in-batch deduplication

t3hw opened the pull request:
Introduce Delta Writer functionality for both unpartitioned and partitioned tables, enabling CDC and upsert modes. Enhance configuration options for CDC fields, upsert mode, and DV usage.

Inspired by #12070
Resolves #10842

@bryanck 

edit:
"using DVs for CDC" - DVs only help for in-batch deduplication.
Out of batch deletes/updates fall back to equality deletes.

Partitioning the table is highly recommended, periodically compacting the table when using CDC mode is mandatory.

t3hw committed:
Iceberg Kafka Connect, DV Mode - with and without Upsert

bryanck reviewed (changes requested):
Thanks for the PR.

As you may know, the [original](https://github.com/databricks/iceberg-kafka-connect) (non-Apache) sink had delta writer support that relied on equality deletes. When the sink was contributed to this project, the community decided it was best to remove that functionality, as using it can result in severely degraded performance. This can lead to poor user experience, reflect badly on the project, and increase support-related questions. Also, there are alternative solutions that don't have the same issues.

We should revisit those discussions and resolve those concerns before we proceed with this.

t3hw commented:
For what it's worth, I tested it under a modest CDC load, and it seems to be working fine.
I can make another commit that locks the CDC mode to use DV only, but that would make the connector incompatible with v2 tables.
Regardless, the exception thrown when the table is a v2 table makes it clear that using DV mode is what causes the failure. At this point, users can decide for themselves how they want to proceed.

t3hw committed:
set useDV->true by default.

t3hw committed:
update readme

t3hw committed:
Merge branch 'apache:main' into cdc-support-with-DV

nalawadechinmay commented:
@bryanck @t3hw Great to see the progress here! Our team is looking to use this connector for a project that requires iceberg sink connector upsert/CDC support. Is there any guidance on whether the implementation part of this PR will be accommodated PR in a later release? Any insight into the timeline or priority would be very helpful for our planning."

hladush commented on kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/data/BaseDeltaWriter.java:
```diff
@@ -0,0 +1,124 @@
+  @Override
+  public void write(Record row) throws IOException {
+
+    Operation op = Operation.fromString(row.getField(cdcField).toString());
```
it will not work with internal fields `iceberg.tables.cdc-field: _cdc.op` 
I think it will be nice to support internal fields because 90%+ engineers will use values from SMT DebeziumTransform

sagarm-traveloka replied:
@t3hw - When are you plan to merge it? For the time being, I have fixed it using below code in BaseDeltaWriter.java - 

```
Operation op =
        Operation.fromString(((Record) row.getField("_cdc")).getField("op").toString());
```

hladush commented:
@bryanck could you review this and, if you’re comfortable, approve?

We’re seeing growing interest from teams that manage their own Kafka infrastructure to use this sink. While our Data Platform teams can achieve similar outcomes with Spark Streaming or the older Iceberg sink (https://github.com/databricks/iceberg-kafka-connect), I believe the current Iceberg sink simplifies day-to-day work for a lot of acquisition teams 
You noted: `This can lead to poor user experience, reflect badly on the project, and increase support-related questions. Also, there are alternative solutions that don't have the same issues.` It's a valid point and totally makes  sense. In my view, the main gap is documentation - something we can fix. Strong docs and examples should reduce support load significantly. The alternatives either require additional infrastructure (e.g., Flink) or are less straightforward to operate for teams (Spark).
If you’re open to it, I’ll add a docs package (setup, config, examples, and a troubleshooting guide) as part of this change. Happy to discuss.

bryanck commented:
@hladush This is something the Iceberg community has decided, including PMC members, so we'd need to get the community on board in order to proceed. You can raise this as a topic on the dev list (again) or in a community sync if you want.

t3hw committed:
Change the fromString translation method to align it more closely to the CdcConstants from the SMT classes.

t3hw commented on kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/data/BaseDeltaWriter.java:
```diff
@@ -0,0 +1,124 @@
+    RowDataDeltaWriter writer = route(row);
+
+    switch (op) {
+      case INSERT:
```
changed to 
``` Java
 public static Operation fromString(String shortString) {
    switch (shortString.toUpperCase(Locale.ROOT)) {
      case "U":
        return UPDATE;
      case "D":
        return DELETE;
      default:
        return INSERT;
    }
  }
```
To align more closely to the SMT CdcConstants

hladush commented:
@bryanck thanks a lot for your response, could you please share with me any doc how to do that ?

supalp commented:
@hladush - thanks for the voice of support for this change. We would love to partner and add our perspective on why this is an important change for consideration by PMV and getting this added in dev list as suggested vt @bryanck

t3hw commented:
added a small clarification to the PR description:

> "using DVs for CDC" - DVs only help for in-batch deduplication.
Out of batch deletes/updates fall back to equality deletes.
> Partitioning the table is highly recommended, periodically compacting the table when using CDC mode is mandatory.

t3hw commented:
Pasting my comment from another discussion about this PR:

Original commenter said:
The primary issue with using DVs for this is identifying the position of a deleted row in the dataset.  If receive a delete record for an existing row, you need to scan the table to locate the file+position in order to create the DV.  That results is lots of scan behavior and data materialization on the worker, which really isn't practical/scalable.
That issue is the main issue that equality deletes address, but equality deletes also introduce significant maintenance burden that isn't available in KC natively.  This means in order to use this approach with KC, you also need Spark or equivalent to aggressively maintain the deletes.
Based on the PR, I'm not clear how these issues are addressed

And the response:
My implementation of BaseDeltaWriter inherits the insertedRowMap from the BaseEqualityDeltaWriter class. It deletes using the DV Writer by default, while falling back to Equality Deletes if the key was not found. The map is cleared when the batch is closed and the file gets written.
I believe RisingWave is using a similar approach.

If this gets accepted the documentation should be updated to let users know that compaction is highly recommended.
The implementation includes `PartitionedDeltaWriter` which routes records to partition-specific writers. This means equality deletes are scoped to their partition - query engines only need to scan files within that partition, not the entire table. For well-partitioned tables (e.g., by date), this significantly reduces the read-time cost of equality deletes.
Another approach would be to keep a separate index of the id-field->parquet file, but running compactions is required regardless, so this is kind of a wasted effort  as it would need to be replicated, persisted, or managed outside of kafka connect, and it would lose its state and have to be re-built once a compaction is triggered.
I agree the PR description was misleading about "using DVs for CDC" - DVs only help for in-batch deduplication. I'll update the description to be clearer.

Given that equality deletes + compaction is how Flink handles CDC (and is the industry standard), would this approach be acceptable if we:
1. Update the PR description to accurately describe the behavior
2. Document that periodic compaction is required for production use
3. Recommend proper table partitioning to minimize equality delete scan overhead
4. Provide example compaction commands in the docs

t3hw committed:
Merge remote-tracking branch 'origin/main' into cdc-support-with-DV

t3hw committed:
Add support for nested cdc fields

t3hw committed:
spotless apply

t3hw committed:
Remove the useDv parameter in favor of automatic detection of table format-version

t3hw committed:
Merge remote-tracking branch 'origin/main' into cdc-support-with-DV

t3hw commented:
> Whether it will merge the multiple records with one id in the one commit interval? In my testing when the job running for some hours there are some duplicate records when there multiple records with one id in the one commit interval by using databricks kafka iceberg connector.

it should

rainerschamm commented:
> @t3hw is it possible to put your code into a new repo ? I want to compile it and to test the cdc upsert

Heya, I merged this into 1.9.2 (and 1.10.1) manually and its works great for me.

You can checkout 1.9.2 and git apply the following:

[cdc-upsert-1.9.2.patch](https://github.com/user-attachments/files/25290264/cdc-upsert-1.9.2.patch)

t3hw commented:
> Will this PR be included in the next release?

I have started a thread on the Kafka Connect Iceberg slack channel about a month ago but it didn't gain much traction. You are more than welcome to assist in the request to merge this feature.

Currently the only option is to compile the branch yourself and drop it into the Kafka Connect plugin folder.

rainerschamm commented:
> > > @t3hw @rainerschamm In my testing there are still duplicated records if oner records are updated frequently. My commit time is 3 minutes. Below are two records updated within one minute and are only two duplicated records in the table. There are 434192 records in the table with 434191 distinct id records.
> > > ## updated_at
> > > 2026-03-05 03:37:58.685000 2026-03-05 03:37:59.076000
> > 
> > 
> > Hmm, we have not seen any duplicates yet in our tests but we only tested it in this setup:
> > 
> > * no partitioning
> > * merge-on-read
> > * 5 minute commit
> > 
> > ... iceberg.tables.auto-create-props.write.delete.mode: merge-on-read iceberg.tables.auto-create-props.write.merge.mode: merge-on-read iceberg.tables.auto-create-props.write.update.mode: merge-on-read ...
> > Also we make sure all identifier fields are strictly non-null in the resulting iceberg table schema.
> 
> @rainerschamm Do you mind share the complete sink properties? Let me check if anything wrong with my config. It is not always have duplicates and only like add a couple of duplicate records everyday. Have no idea how to troubleshoot it.

I can't provide the aws s3tables ones since they are for a customer, but here are the ones we use for demos.

I hit these tables quite hard with random inserts and updates in postgres and they generated no duplicates in iceberg.

[debezium-source-connector.yaml](https://github.com/user-attachments/files/25790651/debezium-source-connector.yaml)
[iceberg-sink-connector.yaml](https://github.com/user-attachments/files/25790652/iceberg-sink-connector.yaml)

t3hw committed:
fix: emit equality delete before INSERT in upsert mode to prevent
cross-batch duplicates

In upsert mode, INSERT did not call `deleteKey()` before `write()`,
unlike Flink's `BaseDeltaTaskWriter` (established in PR #2863, refined
in PR #4364). This caused a "shielding" bug: when INSERT and UPDATE
for the same key arrived in the same batch, INSERT's `write()`
populated `insertedRowMap`, causing UPDATE's `deleteKey()` to take the
position-delete path instead of emitting the equality delete needed to
remove the prior batch's row.

The fix adds `deleteKey(keyProjection.wrap(row))` before `write()` for
INSERT when upsert is enabled - matching Flink's behavior. This
produces an equality delete for every INSERT in upsert mode, even for
genuinely new keys (no-op at read time, resolved by compaction). This
is the same trade-off Flink makes.

Adds table-level integration tests (`TestCDCDeltaWriterTableLevel`)
that commit to real Iceberg tables and read back row data, following
the Flink `TestDeltaTaskWriter` pattern. These tests verify that
equality deletes actually mask prior-batch rows - something the
existing mock-based unit tests could not validate.

Also updates documentation with a delete behavior matrix, exactly-once
semantics explanation, and operational guidance for CDC mode.

t3hw committed:
update Kafka Connect documentation for CDC mode and delete behavior

t3hw commented:
> > Regarding the duplications, @koodin9 wrote a wonderful comment on my other PR (#15651) explaining the issue, and why this fix can lead to a loss off data. More work must be done there before it is safe for use in production. In the mean time, it is possible to use the debezium ts_ms field to deduplicate the rows in the query level.
> 
> @t3hw In my testing, it is actually missing some data

@jerryzhujing I have pushed the [cdc-and-recovery-fix](https://github.com/t3hw/iceberg/tree/cdc-and-recovery-fix) branch with a newer version of the duplicates fix, ill be running some more tests testing it later on week, but its available if you'd like to test it as well

vitaly-sinev commented:
This pull request is very important for us, and we believe it would bring significant value to real-world Kafka Connect and CDC use cases.

We are very interested in helping move it forward and are ready to assist with testing, validation, and any additional work needed to support the review process. If there are specific concerns or next steps that would help unblock progress, we would be glad to contribute.

tburch commented:
I've been using the changes from the PR at moderate volume, and everything is working. I'm also happy to help with any validation or additional work.

juanluhidalgo commented:
What is preventing the merge of this PR since there are people already using it?

rainerschamm commented:
Please can we merge we have been testing this for months now and its working fine for us.

jerryzhujing commented:
Snowflake does not support to read Iceberg tables which using  Equality deletes. Is there any way to fix it? Snowflake supports position deletes and deletion vectors for V3.

ashokcoupa commented:
Hi @bryanck, could we please get some visibility on the status of this PR and whether there are any blockers for merging it?

We have tested this change extensively on our side and it has been working reliably across multiple scenarios. From our perspective, this enhancement adds significant value to the library, especially for upsert-related use cases that several users seem to be facing.

If there are specific concerns or design considerations preventing the merge, it would be very helpful if they could be shared so we can address them. Otherwise, we would strongly appreciate moving this PR forward.

bryanck commented:
@ashokcoupa I left a comment above, in case you missed it, thanks.

juliomaroto commented:
Hey @jerryzhujing ,

I've been dealing with the same issue recenty: when it comes to Snowflake compatibility with Equality Deletes into Iceberg External Tables integration on delta writes. 

I'd say the Snowflake incompatibility with Iceberg Equality Deletes is rather a problem of Snowflake, than a Community Issue (if you have a look to other equivalent computing engines, most of them support equality deletes, e.g. Trino, Athena, or StarRocksDB), as you can see on the Snowflake's public docs, where they state that they only support Iceberg Positional Deletes for External Iceberg Tables Integration/Iceberg Catalog Integration. However, It would be nice/helpful if the connector could optionally include switching from equality deletes to positional deletes, which are supported by Snowflake. 

As per my recent investigations, not even Flink supports equality deletes on Iceberg delta writes, as performance-wise it would be subpar.

Anyway, I agree that optionally including it in the connector would be a great advantage for the community, and extending the adoption of Iceberg, as Snowflake has a wide presence in the data domain in the corporate, and its b2b users base is massive.

Unfortunately I don't think Snowflake will adopt soon iceberg equality deletes, maybe because for now they don't treat Iceberg as a first citizen integration (they rather inviting users to move their OLTP to their new PostgreSQL offering within Snowflake, which gives direct access to the OLTP data generated into their PG to Snowflake), which forces to us the snowflake users to use custom CDC solutions that implement positional deletes if we want to integrate with Snowflake our OLTPs, instead of community-wide OSS solutions, like either this Sink Connector for Kafka, or the Flink one.

That is said, not sure if implementing positional deletes in this Sink Connector, even optionally, is going to be accepted by the maintainers, as it's undrecommended for streaming use cases requiring deletions, such as CDC due to performance issues, as far as I understood when I reasearched about this topic in deep.

rainerschamm commented:
> As per my recent investigations, not even Flink supports equality deletes on Iceberg delta writes, as performance-wise it would be subpar.
> 

The Flink Sink supports upsert mode perfectly:

https://github.com/apache/iceberg/blob/main/flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/sink/BaseDeltaTaskWriter.java

Also, we have been running a manually patched iceberg connect sink with good performance (by extending the same class: org.apache.iceberg.io.BaseTaskWriter in the core iceberg library).

I can not understand how one team in the Iceberg community accepts that "org.apache.iceberg.io.BaseTaskWriter" is behaving well and another team does not.

Regards
Rainer

pvary commented:
The Flink sink is technically functional, but working with tables that contain many equality delete files (for reads and compaction) is challenging and often prohibitively slow and expensive.

There is ongoing work to eliminate the need for equality deletes by moving Flink to a model that relies solely on delete vectors. If this effort is successful, we plan to revisit the role of equality deletes and may consider removing them from the Iceberg specification altogether.

Given this direction, I would recommend avoiding the introduction of new use cases that depend on equality deletes.

t3hw commented:
@pvary wouldn't that require the maintenance of some sort of a per-file index or some primary key equivalent?

pvary commented:
> @pvary wouldn't that require the maintenance of some sort of a per-file index or some primary key equivalent?

Yes. The index proposal is also ongoing, and the Flink solution could be tested in the meantime by reading the initial index data by scanning the original table and projecting only the relevant columns. See: #15996

BenjaminMcC commented:
> @ashokcoupa I left a comment above, in case you missed it, thanks.

@bryanck has your comment already not been addressed up [here](https://github.com/apache/iceberg/pull/14797#issuecomment-3627477095), with a question asked back? 

Im confused as to what the actual hold up is now, there's a lot of community requests and support behind this PR now and its been languishing in review since December.

rainerschamm commented:
> > @ashokcoupa I left a comment above, in case you missed it, thanks.
> 
> @bryanck has your comment already not been addressed up [here](https://github.com/apache/iceberg/pull/14797#issuecomment-3627477095), with a question asked back?
> 
> Im confused as to what the actual hold up is now, there's a lot of community requests and support behind this PR now and its been languishing in review since December.

It looks like the community is working on a better solution at the moment.

In the meantime here is a patch file, based one the original PR, that applies cleanly to 1.10.1 (and 1.9.2) and has been working for us in a CDC Debezium prod environment for a few months now (+200 tables, some with billions of rows) no performance issues and no duplicates. At least Iceberg sink is always quicker than the Debezium source.

[cdc-upsert-1.9.2.patch](https://github.com/user-attachments/files/29325797/cdc-upsert-1.9.2.patch)

juliomaroto commented:
Hey @rainerschamm . I've used recently the CDC upsert fix from @t3hw , which worked for me, but I found it produces equality deletes, instead of positional, which is incompatible with Snowflake (so unfortunately couldn't implement it in-house). 

Would you know if it would be possible to implement positional deletes into the connector? Would be a game changer to drive Iceberg adoption further, as Snowflake product team does not contemplate accepting equality deletes' reads on their roadmap.

Snowflake user-base is just huge, and that would help with the adtoption of Iceberg usage across the industry. Not sure if it's possible to implement positional deletes, as an optional feature/flag (maybe by default equality deletes?).

t3hw commented:
> Hey @rainerschamm . I've used recently the CDC upsert fix from @t3hw , which worked for me, but I found it produces equality deletes, instead of positional, which is incompatible with Snowflake (so unfortunately couldn't implement it in-house).
> 
> Would you know if it would be possible to implement positional deletes into the connector? Would be a game changer to drive Iceberg adoption further, as Snowflake product team does not contemplate accepting equality deletes' reads on their roadmap.
> 
> Snowflake user-base is just huge, and that would help with the adtoption of Iceberg usage across the industry. Not sure if it's possible to implement positional deletes, as an optional feature/flag (maybe by default equality deletes?).

I think it is not possible with the current tools, as it would require the writer knowing the position of the record in the previous snapshot.
The DV is used of an in-batch de-duplication when the writer holds the positional data of the rows it just wrote in memory, but it is lost when the writer commits the snapshot, so out of batch records fall back to Equality Deletes.

see the comments by @pvary above, it seems that there is some work being done on deprecating Equality Deletes and adding a PK/Indexing mechanism. 
Ill keep following that work and perhaps merge it to this branch once it becomes available. 
In the meantime, the solution for kafka connect would be to fork and merge this branch while actively compacting the table.

juliomaroto commented:
Thank you for your quick response @t3hw . As I see you are an active contributor, and this question might help others struggling with the lack of positional deletes compatibility into the kafka connector (presumably Snowflake users). May I ask if there's any production alternative nowadays within Iceberg world which supports positional deletes? Does the Flink connector support positional deletes? I think this would help many Iceberg users looking to adopt a native iceberg solution using kafka + debezium as a source, instead of custom solutions written in Spark Streaming.

t3hw commented:
@juliomaroto Im not sure if there is. If you find one, let me know. i'd love to take a look at how its done

rgizsilva commented:
Hi everyone
iceberg.tables.cdc-field / iceberg.tables.upsert-mode-enabled — still valid in v.1.10.2?
I checked IcebergSinkConfig.java on the apache-iceberg-1.10.2 tag and neither iceberg.tables.cdc-field nor iceberg.tables.upsert-mode-enabled appear in newConfigDef(). Only iceberg.tables.default-id-columns / iceberg.table.<table>.id-columns exist.
But several examples (including configs referenced in #15351) use both properties as if they were valid for org.apache.iceberg.connect.IcebergSinkConnector. I suspect these are leftovers from the old Tabular fork's ConfigDef and were never ported to the ASF connector.
Can someone confirm: is native upsert mode supported at all in the current sink (any property name), or is MERGE INTO downstream still the recommended approach? Unrecognized properties seem to fail silently, which made this confusing to debug.

rainerschamm commented:
> Hi everyone iceberg.tables.cdc-field / iceberg.tables.upsert-mode-enabled — still valid in v.1.10.2? I checked IcebergSinkConfig.java on the apache-iceberg-1.10.2 tag and neither iceberg.tables.cdc-field nor iceberg.tables.upsert-mode-enabled appear in newConfigDef(). Only iceberg.tables.default-id-columns / iceberg.table.
> .id-columns exist.
> But several examples (including configs referenced in #15351) use both properties as if they were valid for org.apache.iceberg.connect.IcebergSinkConnector. I suspect these are leftovers from the old Tabular fork's ConfigDef and were never ported to the ASF connector.
> Can someone confirm: is native upsert mode supported at all in the current sink (any property name), or is MERGE INTO downstream still the recommended approach? Unrecognized properties seem to fail silently, which made this confusing to debug.

No, they are not.
 
They would have been introduced by this PR, but that's not likely to happen at this stage:

```
+  private static final String TABLES_CDC_FIELD_PROP = "iceberg.tables.cdc-field";
+  private static final String TABLES_UPSERT_MODE_ENABLED_PROP =
+      "iceberg.tables.upsert-mode-enabled";
```

t3hw committed:
Merge branch 'main' into cdc-support-with-DV

kinolaev commented:
Slightly off-topic, but since this PR is about the Kafka Connect sink, I built a Flink-based alternative: https://github.com/kinolaev/iceberg-flink-kafka.

[Flink Dynamic Iceberg Sink](https://iceberg.apache.org/docs/latest/flink-writes/#flink-dynamic-iceberg-sink) supports upsert mode, and `iceberg-flink` has an Avro-to-Iceberg mapper. In my project I just wired them together with the [Flink Kafka Connector](https://nightlies.apache.org/flink/flink-docs-stable/docs/connectors/datastream/kafka/). The project supports most of the configuration properties from Iceberg Kafka Connect to ease migration. Additionally, it has a full `compose` environment with `postgres` and `debezium-server` to see it in action. So if you're looking to migrate to Flink Dynamic Iceberg Sink and Avro is a valid serialization option for you, you're more than welcome to test out and use the project.

The Flink sink allows automatic execution of maintenance tasks after data is committed to the table, and recently the [ConvertEqualityDeletes procedure](https://iceberg.apache.org/docs/nightly/flink-maintenance/#convertequalitydeletes) has been merged into the `main` branch. Combined, these two make Flink the most promising solution for CDC scenarios, in my opinion. I don't know if the coordinated maintenance is already integrated with Flink Dynamic Iceberg Sink, but I plan to look into it and provide a way to configure it in my project.
